[Cocoa] Super gigantic images should not be decoded with their natural sizes

shallawa opened the pull request:
#### 37c6fddf3f55237ad288e7f977063f38ea00dea3
<pre>
[Cocoa] Super gigantic images should not be decoded with their natural sizes
<a href="https://bugs.webkit.org/show_bug.cgi?id=259960">https://bugs.webkit.org/show_bug.cgi?id=259960</a>
<a href="https://rdar.apple.com/113305719">rdar://113305719</a>

Reviewed by NOBODY (OOPS!).

(1) For non-scaled image display, asynchronous image decoding should be disabled
to avoid decoding the whole image. Sub-images should be used to display the image
on different tiles. The size of the sub-image is the mapping of the destination
rectangle in the image rectangle clipped by the GraphicsContext.bounds().

(2) DecodingOptions will be refactored to handle the sub-image decoding case.
Instead of having an optional sizeForDrawing, a variant of { monostate, IntRect,
IntSize } will be used. This variant will be named `DecodingGeometry`.

Here is the meaning of the six possible case of DecodingOptions:

    DecodingMode    DecodingGeometry    Image decoding

    Synchronous     std::monostate      Synchronous full image
    Synchronous     IntSize             N/A
    Synchronous     IntRect             Synchronous sub-image
    Asynchronous    std::monostate      Thumbnail full image
    Asynchronous    IntSize             Thumbnail sizeForDrawing
    Asynchronous    IntRect             Not implemented yet

Deciding the sub-image decoding has to happen at the painting time because it can
used only if image sub-sampling is disabled.

(3) An optional `subimageRect` will be added to NativeImage. This member will be
set when a sub-image is decoded. It gives the rectangle of the sub-image in the
rectangle if the PlatformImage.

* LayoutTests/fast/images/resources/green-20000x20000.jpg: Added.
* LayoutTests/fast/images/super-gigantic-image-subsampled-natural-size-expected.html: Added.
* LayoutTests/fast/images/super-gigantic-image-subsampled-natural-size.html: Added.
* Source/WebCore/platform/graphics/BitmapImage.cpp:
(WebCore::BitmapImage::draw):
* Source/WebCore/platform/graphics/BitmapImage.h:
* Source/WebCore/platform/graphics/BitmapImageSource.cpp:
(WebCore::BitmapImageSource::startAnimation):
(WebCore::BitmapImageSource::isGiganticForDecoding const):
(WebCore::BitmapImageSource::nativeImageAtIndexCacheIfNeeded):
(WebCore::BitmapImageSource::decodingOptionsForDrawing const):
(WebCore::BitmapImageSource::currentNativeImageForDrawing):
* Source/WebCore/platform/graphics/BitmapImageSource.h:
* Source/WebCore/platform/graphics/DecodingOptions.h:
(WebCore::DecodingOptions::DecodingOptions):
(WebCore::DecodingOptions::decodingGeometry const):
(WebCore::DecodingOptions::sizeForDrawing const):
(WebCore::DecodingOptions::rectForDrawing const):
(WebCore::DecodingOptions::hasSizeForDrawing const):
(WebCore::DecodingOptions::hasRectForDrawing const):
(WebCore::DecodingOptions::hasFullSize const):
(WebCore::DecodingOptions::isCompatibleWith const):
* Source/WebCore/platform/graphics/ImageDecoder.h:
(WebCore::ImageDecoder::createFrameSubimageAtIndex):
* Source/WebCore/platform/graphics/ImageFrame.cpp:
(WebCore::ImageFrame::hasFullSizeNativeImage const): Deleted.
* Source/WebCore/platform/graphics/ImageFrame.h:
(WebCore::ImageFrame::hasNativeImage):
(WebCore::ImageFrame::hasFullSizeNativeImage): Deleted.
* Source/WebCore/platform/graphics/ImageSource.cpp:
(WebCore::ImageSource::primaryNativeImageForDrawing):
(WebCore::ImageSource::currentNativeImageForDrawing):
* Source/WebCore/platform/graphics/ImageSource.h:
(WebCore::ImageSource::isGiganticForDecoding const):
* Source/WebCore/platform/graphics/NativeImage.cpp:
(WebCore::NativeImage::create):
(WebCore::NativeImage::createTransient):
(WebCore::NativeImage::NativeImage):
* Source/WebCore/platform/graphics/NativeImage.h:
* Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp:
(WebCore::ImageDecoderCG::createFrameSubimageAtIndex):
* Source/WebCore/platform/graphics/cg/ImageDecoderCG.h:
* Source/WebCore/platform/graphics/cg/NativeImageCG.cpp:
(WebCore::NativeImage::create):
(WebCore::NativeImage::createTransient):
* Source/WebKit/WebProcess/WebCoreSupport/ShareableBitmapUtilities.cpp:
(WebKit::createShareableBitmap):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/37c6fddf3f55237ad288e7f977063f38ea00dea3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/48600 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/27811 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/51561 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/51287 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/44665 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/50905 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/33748 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/25341 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/39745 "Passed tests") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/49182 "Build is in progress. Recent messages:OS: Monterey (12.6.8), Xcode: 13.4.1; Skipping applying patch since patch_id isn't provided; Passed webkitperl tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/25476 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/41931 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20842 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/22954 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/43120 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/6656 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/44893 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/43594 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/53195 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/23647 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/19946 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/47037 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/24913 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/42136 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/45965 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/25717 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/24634 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/platform/graphics/BitmapImage.h:
```diff
@@ -158,6 +158,11 @@ class BitmapImage final : public Image {
 
+    // Async image decoding should be allowed between these two sizes depending on
+    // whether it is animated or static image.
+    static constexpr auto maxImageSizeForSyncDecoding = FloatSize { 400, 400 };
```
Might be easier to understand as `minImageSizeForAsyncDecoding`

smfr commented on Source/WebCore/platform/graphics/ImageSource.cpp:
```diff
@@ -337,8 +337,19 @@ bool ImageSource::canUseAsyncDecoding()
+    float frameSizeInPixels = size().area();
+
+    auto maxImageSizeForSyncDecoding = BitmapImage::maxImageSizeForSyncDecoding;
+    if (frameSizeInPixels < (frameCount() > 1 ? maxImageSizeForSyncDecoding.area() / 8 : maxImageSizeForSyncDecoding.area()))
```
The old "animated image" ratio was 1/5, not 1/8 like here. Any reason for changing it?

smfr commented on Source/WebCore/platform/graphics/ImageSource.cpp:
```diff
@@ -337,8 +337,19 @@ bool ImageSource::canUseAsyncDecoding()
+
+    // Super gigantic images have to be drawn as sub-images or sub-sampled.
+    auto maxImageSizeForWholeDecoding = BitmapImage::maxImageSizeForWholeDecoding;
+    if (frameSizeInPixels > (frameCount() > 1 ? maxImageSizeForWholeDecoding.area() / 8 : maxImageSizeForWholeDecoding.area()))
```
Some ` LOG_WITH_STREAM(Images...)` here might be useful to understand why we hit or not the async path for a given image.

smfr commented on Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp:
```diff
@@ -263,6 +263,13 @@ void Recorder::drawNativeImageInternal(NativeImage& image, const FloatSize& imag
+{
+    appendStateChangeItemIfNecessary();
+    recordResourceUse(image);
+    recordDrawNativeSubImage(image.renderingResourceIdentifier(), imageSize, destRect, srcRect, options);
```
Do we need to be careful with `image.renderingResourceIdentifier()` here? Might we end up trying to cache both a whole image and subimage with the same identifier? I'd think that we may need a unique identifier for a subsampled or subimage instance of a given image.

smfr reviewed (commented):
I need to be sure that the resource cache behavior is correct before approving.

shallawa committed:
[Cocoa] Super gigantic images should not be decoded with their natural sizes
https://bugs.webkit.org/show_bug.cgi?id=259960
rdar://113305719

Reviewed by NOBODY (OOPS!).

(1) For non-scaled image display, asynchronous image decoding should be disabled
to avoid decoding the whole image. Sub-images should be used to display the image
on different tiles. The size of the sub-image is the mapping of the destination
rectangle in the image rectangle clipped by the GraphicsContext.bounds().

(2) DecodingOptions will be refactored to handle the sub-image decoding case.
Instead of having an optional sizeForDrawing, a variant of { monostate, IntRect,
IntSize } will be used. This variant will be named `DecodingGeometry`.

Here is the meaning of the six possible case of DecodingOptions:

    DecodingMode    DecodingGeometry    Image decoding

    Synchronous     std::monostate      Synchronous full image
    Synchronous     IntSize             N/A
    Synchronous     IntRect             Synchronous sub-image
    Asynchronous    std::monostate      Thumbnail full image
    Asynchronous    IntSize             Thumbnail sizeForDrawing
    Asynchronous    IntRect             Not implemented yet

Deciding the sub-image decoding has to happen at the painting time because it can
used only if image sub-sampling is disabled.

(3) An optional `subimageRect` will be added to NativeImage. This member will be
set when a sub-image is decoded. It gives the rectangle of the sub-image in the
rectangle if the PlatformImage.

* LayoutTests/fast/images/resources/green-20000x20000.jpg: Added.
* LayoutTests/fast/images/super-gigantic-image-subsampled-natural-size-expected.html: Added.
* LayoutTests/fast/images/super-gigantic-image-subsampled-natural-size.html: Added.
* Source/WebCore/platform/graphics/BitmapImage.cpp:
(WebCore::BitmapImage::draw):
* Source/WebCore/platform/graphics/BitmapImage.h:
* Source/WebCore/platform/graphics/BitmapImageSource.cpp:
(WebCore::BitmapImageSource::startAnimation):
(WebCore::BitmapImageSource::isGiganticForDecoding const):
(WebCore::BitmapImageSource::nativeImageAtIndexCacheIfNeeded):
(WebCore::BitmapImageSource::decodingOptionsForDrawing const):
(WebCore::BitmapImageSource::currentNativeImageForDrawing):
* Source/WebCore/platform/graphics/BitmapImageSource.h:
* Source/WebCore/platform/graphics/DecodingOptions.h:
(WebCore::DecodingOptions::DecodingOptions):
(WebCore::DecodingOptions::decodingGeometry const):
(WebCore::DecodingOptions::sizeForDrawing const):
(WebCore::DecodingOptions::rectForDrawing const):
(WebCore::DecodingOptions::hasSizeForDrawing const):
(WebCore::DecodingOptions::hasRectForDrawing const):
(WebCore::DecodingOptions::hasFullSize const):
(WebCore::DecodingOptions::isCompatibleWith const):
* Source/WebCore/platform/graphics/ImageDecoder.h:
(WebCore::ImageDecoder::createFrameSubimageAtIndex):
* Source/WebCore/platform/graphics/ImageFrame.cpp:
(WebCore::ImageFrame::hasFullSizeNativeImage const): Deleted.
* Source/WebCore/platform/graphics/ImageFrame.h:
(WebCore::ImageFrame::hasNativeImage):
(WebCore::ImageFrame::hasFullSizeNativeImage): Deleted.
* Source/WebCore/platform/graphics/ImageSource.cpp:
(WebCore::ImageSource::primaryNativeImageForDrawing):
(WebCore::ImageSource::currentNativeImageForDrawing):
* Source/WebCore/platform/graphics/ImageSource.h:
(WebCore::ImageSource::isGiganticForDecoding const):
* Source/WebCore/platform/graphics/NativeImage.cpp:
(WebCore::NativeImage::create):
(WebCore::NativeImage::createTransient):
(WebCore::NativeImage::NativeImage):
* Source/WebCore/platform/graphics/NativeImage.h:
* Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp:
(WebCore::ImageDecoderCG::createFrameSubimageAtIndex):
* Source/WebCore/platform/graphics/cg/ImageDecoderCG.h:
* Source/WebCore/platform/graphics/cg/NativeImageCG.cpp:
(WebCore::NativeImage::create):
(WebCore::NativeImage::createTransient):
* Source/WebKit/WebProcess/WebCoreSupport/ShareableBitmapUtilities.cpp:
(WebKit::createShareableBitmap):
